### PR TITLE
[Feature](bangc-ops): add points_in_boxes op

### PR DESCRIPTION
**Pull request by caixianzhang**
Thanks for your contribution and we appreciate it a lot. 

## 1. Motivation

Add points_in_boxes op.

## 2. Modification

Add points_in_boxes op.

#### 3.1.1 Accuracy Acceptance Standard

For static threshold standard details, see: [MLU-OPS Accuracy Acceptance Standard](https://github.com/Cambricon/mlu-ops/blob/master/docs/MLU-OPS-Accuracy-Acceptance-Standard.md).

- [x] diff3: 0

#### 3.1.2 Operator Scheme checklist

|     No.        |                 Details              |            Check Results             |
|----------------|--------------------------------------|--------------------------------------|
|        1       |Supported hardware                    |             MLU370<br>MLU590         |
|        2       |Job types                             |          block      |
|        3       |Layouts                               |          ARRAY       |
|        4       |Whether multi-dimensions are supported|             No                         |
|        5       |Whether element zero is supported     |           No                           |
|        6       |Data type(half/float)                 |          float          |
|        7       |Whether there is size limit           |    On MLU370, the number of boxes cannot exceed 23404; on MLU590, the number of boxes cannot exceed 14042.                                  |

#### 3.1.3 New Feature Test

If you have checked the following items, please tick the relevant box.

- [x] Data type test
- [ ] Multi-dimensional tensor test
- [x] Layout test
- [ ] Different size/integer remainder end segment/alignment misalignment test
- [x] Zero dimensional tensor test/zero element test
- [x] stability test
- [x] Multiple platform test
- [x] Gen_case module test
- [x] Nan/INF tests 
- [ ] Bug fix tests
- [x] For memory leak check details, see[GTest-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/GTest-User-Guide-zh.md).
- [x] For code coverage check details, see: [GTest-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/GTest-User-Guide-zh.md).
- [x] For I/O calculation efficiency check details, see: [MLU-OPS Performance Acceptance Standard](https://github.com/Cambricon/mlu-ops/blob/master/docs/MLU-OPS-Performance-Acceptance-Standard.md).

#### 3.1.4 Parameter Check

When a new operator is submitted, the test points are given and the test results are stated.

|                   Test Point                    | Acceptance Standard | Test Result (Error Message) |
| ----------------------------------------------- | --------------------| --------------------------- |
| Whether it conforms to the operator restriction |     Normal error    |                             |
| Whether illegal parameters are passed           |     Normal error    |                             |

### 3.2 Accuracy Test

For the cases used in the New Feature Test section, the features and the number of cases are recorded here. When multiple operations are tested, multiple tables are needed to include details of these operations.

Operation:

|Test Point           | Description                      | Quantity |  Comment |
|----------           |----------------------------------|----------|  --------|
|Data type test       | float                   |     图片     |          |
|Mult-tensor test     |  Only supports 3 dims                 |     see above     |          |
|Layout test          | Only supports Array                |   see above       |          |
|Zero element test    | Yes    |     图片   |          |
|Stability test       |--gtest_repeat=1000<br>--thread=4|    图片    |          |
|Mult-platform test   |MLU370/MLU590                     |     支持MLU370,MLU590    |          |
|Nan/INF test         |Whether to support this test      |  points_in_boxes对nan/inf处理与竞品保持一致    |          |
|Memory leak check    |Test result                       |     测试通过     |          |
|Code coverage check  |Test result                       |   图片   |          |

### 3.3 Performance Test

See [MLU-OPS Performance Acceptance Standard](https://github.com/Cambricon/mlu-ops/blob/master/docs/MLU-OPS-Performance-Acceptance-Standard.md) for details.

Platform：MLU370

|Operation|Mlu_hardware_time(us)|Mlu_interface_time(us)|Mlu_io_efficiency|Mlu_compute_efficiency|Mlu_workwpace_size(Bytes)|Data_type|Shape|
|-------|----|----|-----|----|----|----|-----|
|points_in_boxes| 454  | 13.09  |   0.0314068  |  0.820548  |  -1  |  float  |  points:[1,273651,3], boxes:[1,66,7]  |

Platform：MLU590

|Operation|Mlu_hardware_time(us)|Mlu_interface_time(us)|Mlu_io_efficiency|Mlu_compute_efficiency|Mlu_workwpace_size(Bytes)|Data_type|Shape|
|-------|----|----|----|----|----|----|-----|
|points_in_boxes|  207  |  4.373  | 0.0103324   |  0.799847  |  -1  |  float  |  points:[1,273651,3], boxes:[1,66,7]   |

### 3.4 Summary Analysis

Please give a brief overview here, if you want to note and summarize the content.

### CODE REVIEW

**Review by Devin-D-U — APPROVED** [review]

File: bangc-ops/kernels/points_in_boxes/points_in_boxes_block_kernel.mlu [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+    uint32_t batch_size, uint32_t points_num, uint32_t boxes_num, float *points,
+    float *boxes, int *idx, uint32_t points_batch_offset,
+    uint32_t boxes_batch_offset, uint32_t idx_batch_offset,
+    uint32_t points_deal_num, uint32_t points_deal_offset,
+    uint32_t idx_deal_num) {
+  MLUBlockKernelPointsInBoxes<<<k_dim, k_type, queue>>>(
+      batch_size, points_num, boxes_num, points, boxes, idx,
+      points_batch_offset, boxes_batch_offset, idx_batch_offset,
+      points_deal_num, points_deal_offset, idx_deal_num);
+}
```
[the remark this answers is not in the document]
Comment (caixianzhang) [reply bangc-ops/kernels/points_in_boxes/points_in_boxes_block_kernel.mlu:231]: host 调用只有cncc 能识别，所以 <<<>>> 结构的代码需要写在 .mlu 文件中

**Comment by duzekunKTH** [comment]
PR标题与 commit 格式错误需要修改

**Comment by caixianzhang** [comment]
> PR标题与 commit 格式错误需要修改



> PR标题与 commit 格式错误需要修改

已修改

**Review by alex-xuwenming — APPROVED** [review]

**Comment by alex-xuwenming** [comment]
LGTM

File: bangc-ops/mlu_op.h [hunk 2]
```diff
@@ -10800,6 +10800,65 @@ mluOpMoeDispatchBackwardGate(mluOpHandle_t handle,
...
 
+// Group:PointsInBoxes
+/*!
+ * @brief Detects the first 3D box that each point belongs to in given points cloud data.
+ *
+ * @param[in] handle
+ * Handle to an MLUOP context that is used to manage MLU devices and queues in the
+ * points_in_boxes operation. For detailed information, see ::mluOpHandle_t.
+ * @param[in] points_desc
+ * The descriptor of input tensor \b points, which contains dimension, data type and data layout.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:10811]: 每一行前面都需要有*

File: bangc-ops/mlu_op.h [hunk 3]
```diff
@@ -10800,6 +10800,65 @@ mluOpMoeDispatchBackwardGate(mluOpHandle_t handle,
...
+ * For detailed information, see ::mluOpTensorDescriptor_t.
+ * @param[in] points
+ * Pointer to the MLU memory that stores the \b points tensor.
+ * @param[in] boxes_desc
+ * The descriptor of input tensor \b boxes, which contains dimension, data type and data layout.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:10816]: 每一行前面都需要有*

File: bangc-ops/mlu_op.h [hunk 4]
```diff
@@ -10800,6 +10800,65 @@ mluOpMoeDispatchBackwardGate(mluOpHandle_t handle,
...
+ * @param[in] boxes
+ * Pointer to the MLU memory that stores the \b boxes tensor.
+ * @param[out] points_indices_desc
+ * The descriptor of input tensor \b points_indices, which contains dimension, data type and data layout.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:10820]: 每一行前面都需要有*

**Review by GraceGuanH — COMMENTED** [review]
Please recheck the comments.

**Review by GraceGuanH — APPROVED** [review]

**Review by duzekunKTH — APPROVED** [review]

**Pull request merged**